File: dfs/max_rising_path/max_rising_path.cpp
#include <list>
#include <string>
#include <vector>
using namespace std;

class MaxRisingPath
{ // LeetCode Q.329.
private:
    int maxRisingLen = 1; // Base case.
    vector<vector<int>> longestRisingLens;

    int totalRows;
    int totalCols;
    vector<vector<int>> storedMatrix;

    void dfsMaxRisingPath(int rowIdx, int colIdx)
    {
        list<vector<int>> neighbors;

        if (colIdx < totalCols - 1)
        { // Can go East.
            neighbors.push_back({rowIdx, colIdx + 1});
        }

        if (0 < colIdx)
        { // Can go West.
            neighbors.push_back({rowIdx, colIdx - 1});
        }

        if (rowIdx < totalRows - 1)
        { // Can go South.
            neighbors.push_back({rowIdx + 1, colIdx});
        }

        if (0 < rowIdx)
        { // Can go North.
            neighbors.push_back({rowIdx - 1, colIdx});
        }

        int maxNeighborLen = 0;
        for (auto neighbor : neighbors)
        {
            int nextRowIdx = neighbor[0];
            int nextColIdx = neighbor[1];
            int neighborVal = storedMatrix[nextRowIdx][nextColIdx];

            if (neighborVal > storedMatrix[rowIdx][colIdx])
            {
                if (longestRisingLens[nextRowIdx][nextColIdx] == 0)
                {
                    dfsMaxRisingPath(nextRowIdx, nextColIdx);
                }

                int neighborLen = longestRisingLens[nextRowIdx][nextColIdx];

                if (neighborLen > maxNeighborLen)
                {
                    maxNeighborLen = neighborLen;
                }
            }
        }

        longestRisingLens[rowIdx][colIdx] = 1 + maxNeighborLen;
        if (longestRisingLens[rowIdx][colIdx] > maxRisingLen)
        {
            maxRisingLen = longestRisingLens[rowIdx][colIdx];
        }
    }

public:
    int findLongestRisingPath(vector<vector<int>> &matrix)
    {
        storedMatrix = matrix;
        totalRows = matrix.size();
        totalCols = matrix[0].size();

        // Initialize matrix to 0, denoting all unsearched cells.
        longestRisingLens = vector<vector<int>>(
            totalRows, vector<int>(totalCols, 0));

        for (int rowIdx = 0; rowIdx < totalRows; rowIdx++)
        {
            for (int colIdx = 0; colIdx < totalCols; colIdx++)
            {
                if (longestRisingLens[rowIdx][colIdx] == 0)
                {
                    dfsMaxRisingPath(rowIdx, colIdx);
                }
            }
        }
        return maxRisingLen;
    }
};

File: dfs/max_rising_path/max_rising_path.py

class MaxRisingPath:  # LeetCode Q.329.
    def __init__(self):
        self.matrix = []
        self.total_rows, self.total_cols = 0, 0
        self.longest_rising_lens = []
        self.max_rising_len = 1

    def find_longest_rising_path(self, matrix: list[list[int]]) -> int:
        self.matrix = matrix
        self.total_rows, self.total_cols = len(matrix), len(matrix[0])
        self.longest_rising_lens = [
            [0 for _ in range(self.total_cols)] for _ in range(self.total_rows)
        ]  # 0 denotes unsearched cell.
        self.max_rising_len = 1  # Base case.

        for row_idx in range(self.total_rows):
            for col_idx in range(self.total_cols):
                if self.longest_rising_lens[row_idx][col_idx] == 0:
                    self._dfs_max_rising_path(row_idx, col_idx)

        return self.max_rising_len

    def _dfs_max_rising_path(self, row_idx: int, col_idx: int) -> None:
        neighbors = []
        if col_idx < self.total_cols - 1:  # Can go East.
            neighbors.append((row_idx, col_idx + 1))

        if 0 < col_idx:  # Can go West.
            neighbors.append((row_idx, col_idx - 1))

        if row_idx < self.total_rows - 1:  # Can go South.
            neighbors.append((row_idx + 1, col_idx))

        if 0 < row_idx:  # Can go North.
            neighbors.append((row_idx - 1, col_idx))

        max_neighbor_len = 0
        for next_row_idx, next_col_idx in neighbors:
            neighbor = self.matrix[next_row_idx][next_col_idx]
            if neighbor > self.matrix[row_idx][col_idx]:
                if self.longest_rising_lens[next_row_idx][next_col_idx] == 0:
                    self._dfs_max_rising_path(next_row_idx, next_col_idx)

                neighbor_len = self.longest_rising_lens[next_row_idx][next_col_idx]
                if neighbor_len > max_neighbor_len:
                    max_neighbor_len = neighbor_len

        self.longest_rising_lens[row_idx][col_idx] = 1 + max_neighbor_len
        if self.longest_rising_lens[row_idx][col_idx] > self.max_rising_len:
            self.max_rising_len = self.longest_rising_lens[row_idx][col_idx]
